src/audio/ipcgtw.c

```diff
@@ -0,0 +1,397 @@
+					audio_stream_get_free_bytes(&buf_c->stream));
+			dcache_invalidate_region((__sparse_force void __sparse_cache *)
+						 MAILBOX_HOSTBOX_BASE,
+						 data_size +
+						 offsetof(struct ipc4_ipc_gateway_cmd_data,
+							  payload));
+			audio_stream_copy_bytes_from_linear(in->payload, &buf_c->stream,
+							    data_size);
+			buffer_stream_writeback(buf_c, data_size);
+			comp_update_buffer_produce(buf_c, data_size);
+			out->u.size_consumed = data_size;
+			*reply_payload_size = 4;
+		} else {
+			out->u.size_consumed = 0;
+			*reply_payload_size = 4;
+		}
+		break;
+
+	case IPCGWCMD_FLUSH_DATA:
+		*reply_payload_size = 0;
+		if (buf_c)
+			audio_stream_reset(&buf_c->stream);
+		break;
+
+	default:
+		comp_cl_err(&comp_ipcgtw, "ipcgtw_process_cmd(): unexpected cmd: %u",
+			    (unsigned int)cmd->primary.r.cmd);
+		if (buf_c)
+			buffer_release(buf_c);
+		return -EINVAL;
+	}
+
+	if (buf_c)
+		buffer_release(buf_c);
+	return 0;
+}
+
+static int ipcgtw_copy(struct comp_dev *dev)
+{
+	return 0;
+}
+
+static int ipcgtw_params(struct comp_dev *dev, struct sof_ipc_stream_params *params)
+{
+	struct comp_buffer *buf;
+	struct comp_buffer __sparse_cache *buf_c;
+	int err;
+	struct ipcgtw_data *ipcgtw_data = comp_get_drvdata(dev);
+
+	comp_cl_dbg(&comp_ipcgtw, "ipcgtw_params()");
+
+	buf = get_buffer(dev);
+	if (!buf) {
+		comp_cl_err(&comp_ipcgtw, "ipcgtw_params(): no buffer found");
+		return -EINVAL;
+	}
+
+	/* resize buffer to size specified in IPC gateway config blob */
+	buf_c = buffer_acquire(buf);
+	err = buffer_set_size(buf_c, ipcgtw_data->buf_size);
+	buffer_release(buf_c);
+
+	if (err < 0) {
+		comp_cl_err(&comp_ipcgtw, "ipcgtw_params(): failed to resize buffer to %u bytes",
+			    ipcgtw_data->buf_size);
+		return err;
+	}
+
+	return 0;
+}
+
+static int ipcgtw_trigger(struct comp_dev *dev, int cmd)
+{
+	/* copier calls gateway ops.trigger() without checking if it is NULL or not,
+	 * so this handler is needed mostly to prevent crash.
+	 */
+	int ret = comp_set_state(dev, cmd);
+
+	if (ret < 0)
+		return ret;
+
+	if (ret == COMP_STATUS_STATE_ALREADY_SET)
+		return PPL_STATUS_PATH_STOP;
+
+	return 0;
+}
+
+static int ipcgtw_prepare(struct comp_dev *dev)
+{
+	/* copier calls gateway ops.prepare() without checking if it is NULL or not,
+	 * so this handler is needed mostly to prevent crash.
+	 */
+	int ret = comp_set_state(dev, COMP_TRIGGER_PREPARE);
+
+	if (ret < 0)
+		return ret;
+
+	if (ret == COMP_STATUS_STATE_ALREADY_SET)
+		return PPL_STATUS_PATH_STOP;
+
+	return 0;
+}
+
+static int ipcgtw_reset(struct comp_dev *dev)
+{
+	struct comp_buffer *buf = get_buffer(dev);
+
+	if (buf) {
+		struct comp_buffer __sparse_cache *buf_c = buffer_acquire(buf);
+
+		audio_stream_reset(&buf_c->stream);
+		buffer_release(buf_c);
+	} else {
+		comp_cl_warn(&comp_ipcgtw, "ipcgtw_reset(): no buffer found");
+	}
+
+	comp_set_state(dev, COMP_TRIGGER_RESET);
+
+	return 0;
+}
+
+static const struct comp_driver comp_ipcgtw = {
```

Comment:
pleae don't use the comp_drv ops. Any new module should use the module adapter

Reply:
Yes, but I made IPC gateway similarly to existing host and DAI gateways. They use comp_drv ops API. These are not real components, i.e., those ops do not called from pipeline but from copier itself. So copier_copy() calls endpoint[i]->drv->ops.copy(), same for .params(), .prepare(), .create(), etc. So to use new module adapter API, first, copier, host-zephyr and dai-zephyr (not sure about host-legacy and dai-legacy) have to be modified to use module adapter API. That's why I use existing comp_drv ops API.

Reply:
@ranj063 @serhiy-katsyuba-intel it sounds like we need to update the comp_driver APIs to remove any pipeline helpers and align the docs and APIs to be called as "low level" after all the high level modules are updated to module API.